x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -0,0 +1,298 @@
+            if (Strings.hasText(requestedFormat)) {
+                authnState.put("nameid_format", requestedFormat);
+                // we should not throw an error. Pass this as additional data so that the /saml/init API can
+                // return a SAML response with the appropriate status (3.4.1.1 in the core spec)
+                if (requestedFormat.equals(UNSPECIFIED) == false &&
+                    requestedFormat.equals(sp.getNameIDPolicyFormat()) == false) {
+                    logger.warn(() ->
+                        new ParameterizedMessage("The requested NameID format [{}] doesn't match the allowed NameID format" +
+                            "for this Service Provider is [{}]", requestedFormat, sp.getNameIDPolicyFormat()));
+                    authnState.put("error", "invalid_nameid_policy");
+                }
+            }
+        }
+        return authnState;
+    }
+
+    private void validateAuthnRequest(AuthnRequest authnRequest, SamlServiceProvider sp) {
+        checkDestination(authnRequest);
+        checkAcs(authnRequest, sp);
+    }
+
+    private boolean validateSignature(String samlRequest, String sigAlg, String signature, X509Credential credential,
+                                      @Nullable String relayState) {
+        try {
+            final String queryParam = relayState == null ?
+                "SAMLRequest=" + urlEncode(samlRequest) + "&SigAlg=" + urlEncode(sigAlg) :
+                "SAMLRequest=" + urlEncode(samlRequest) + "&RelayState=" + urlEncode(relayState) + "&SigAlg=" + urlEncode(sigAlg);
+            Signature sig = Signature.getInstance(getJavaAlorithmNameFromUri(sigAlg));
+            sig.initVerify(credential.getEntityCertificate().getPublicKey());
+            sig.update(queryParam.getBytes(StandardCharsets.UTF_8));
+            return sig.verify(Base64.getDecoder().decode(signature));
+        } catch (Exception e) {
+            throw new ElasticsearchSecurityException("Unable to validate signature of authentication request using credentials [{}]",
+                SamlUtils.describeCredentials(Collections.singletonList(credential)), e);
+        }
+    }
+
+    private SamlServiceProvider getSpFromIssuer(Issuer issuer) {
+        if (issuer == null || issuer.getValue() == null) {
+            throw new ElasticsearchSecurityException("SAML authentication request has no issuer");
+        }
+        final String issuerString = issuer.getValue();
+        final SamlServiceProvider serviceProvider = idp.getRegisteredServiceProvider(issuerString);
+        if (null == serviceProvider) {
+            throw new ElasticsearchSecurityException("Service Provider with Entity ID [{}] is not registered with this Identity Provider",
+                issuerString);
+        }
+        return serviceProvider;
+    }
+
+    private void checkDestination(AuthnRequest request) {
+        final String url = idp.getSingleSignOnEndpoint("redirect");
```

Comment:
Can we move this (`"redirect"`) to a constant?

Reply:
Yes, I've already changed it in https://github.com/elastic/elasticsearch/pull/52415 to use `org.opensaml.saml.common.xml.SAMLConstants.SAML2_REDIRECT_BINDING_URI` instead of simple strings, will adjust it here too